node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+// includable when some bitfields are included. And we might have candidates
+// that are not includable when certain bitfields are included.
+//
+// When we're choosing bitfields to include, the rule should be simple:
+// maximize availability. So basically, include all bitfields. And then
+// choose a coherent set of candidates along with that.
+async fn send_inherent_data(
+	relay_parent: Hash,
+	signed_bitfields: Vec<SignedAvailabilityBitfield>,
+	mut backed_candidates: Vec<BackedCandidate>,
+	return_sender: oneshot::Sender<ProvisionerInherentData>,
+	mut from_job: mpsc::Sender<FromJob>,
+) -> Result<(), Error> {
+	let availability_cores = match request_availability_cores(relay_parent, &mut from_job)
+		.await?
+		.await?
+	{
+		Ok(cores) => cores,
+		Err(runtime_err) => {
+			// Don't take down the node on runtime API errors.
+			log::warn!(target: "provisioner", "Encountered a runtime API error: {:?}", runtime_err);
+			return Ok(());
+		}
+	};
+
+	// availability cores indexed by their para_id for ease of use
+	let cores_by_id: HashMap<_, _> = availability_cores
+		.iter()
+		.enumerate()
+		.filter_map(|(idx, core_state)| Some((core_state.para_id()?, (idx, core_state))))
+		.collect();
+
+	// utility
+	let mut parachains_represented = HashSet::new();
+
+	// coherent candidates fulfill these conditions:
+	//
+	// - only one per parachain
+	// - any of:
+	//   - this para is assigned to a `Scheduled` core
+	//   - this para is assigned to an `Occupied` core, and any of:
+	//     - it is `next_up_on_available` and the bitfields we are including, merged with
+	//       the `availability` vec, form 2/3+ of validators
+	//     - it is `next_up_on_time_out` and the bitfields we are including, merged with
+	//       the `availability_ vec, for <2/3 of validators, and `time_out_at` is
+	//       the block we are building.
+	//
+	// postcondition: they are sorted by core index
+	backed_candidates.retain(|candidate| {
+		// only allow the first candidate per parachain
+		let para_id = candidate.candidate.descriptor.para_id;
+		if !parachains_represented.insert(para_id) {
+			return false;
+		}
+
+		let (core_idx, core) = match cores_by_id.get(&para_id) {
+			Some(core) => core,
+			None => return false,
+		};
+
+		match core {
+			CoreState::Free => false,
+			CoreState::Scheduled(_) => true,
+			CoreState::Occupied(occupied) => {
+				if let Some(ref scheduled) = occupied.next_up_on_available {
+					return scheduled.para_id == para_id
+						&& merged_bitfields_are_gte_two_thirds(
+							*core_idx,
+							&signed_bitfields,
+							&occupied.availability,
+						);
+				}
+				if let Some(ref scheduled) = occupied.next_up_on_time_out {
+					return scheduled.para_id == para_id
+						// TODO: this function currently doesn't know about what block we're building
+						//   get it from the runtime? where?
+						&& unimplemented!("occupied.time_out_at == the block we are building")
```

Comment:
This is the last significant unimplemented section.

Reply:
Probably depends on the Chain API to get block number by hash. We would call that with the hash of the relay-parent and then add 1 to get the number of the child.

Reply:
An alternative is to have the `RequestInherentData` pass the number of the child block and assume that it is correct.

Reply:
Is it intentional that `ChainApiMessage` is not a variant of `AllMessages`, or an omission?

I started out thinking that just adding the child block number to `RequestInherentData` would be the best way forward. On exploring it, it seems that getting the block number under construction from within the provisioner's context is un-obvious. It may be possible i.e. by adding trait bounds, but it was unclear. Using the Chain API seems the simpler solution, but it does not actually exist within `AllMessages` for now, and it's unclear whether or not that was on purpose.

Reply:
it's an omission in https://github.com/paritytech/polkadot/pull/1498, it should be integrated with `AllMessages` and `utils` module.

Reply:
Cool, I'll add it. Thanks!